### PR TITLE
ENG-10404 support for start and init cli commands

### PR DESCRIPTION
**Pull request by vtkstef**

### CODE REVIEW

File: lib/python/voltcli/voltdb.d/init.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+                                needs_catalog=False,
+                                supports_live=False,
+                                default_host=False,
+                                safemode_available=False,
+                                supports_daemon=False,
+                                supports_multiple_daemons=False,
+                                check_environment_config=True,
+                                force_voltdb_create=True,
+                                supports_paused=False,
+                                is_legacy_verb=False),
```
Comment (akhanzode) [inline lib/python/voltcli/voltdb.d/init.py]: Can you remove all network options from init ?

File: src/frontend/org/voltcore/messaging/HostCriteria.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    public final static String HOST_COUNT = "hostCount";
+    public final static String NODE_STATE = "nodeState";
+    public final static String ADD_ALLOWED = "addAllowed";
+
+    public final static UUID UNDEFINED = new UUID(0L,0L);
+
+    public static boolean hasCriteria(JSONObject jo) {
+        try {
+            return jo != null
+                && jo.has(IS_PAUSED) && (jo.getBoolean(IS_PAUSED) ? true : true)
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/HostCriteria.java]: This is a long && check any better way?

File: src/frontend/org/voltcore/messaging/HostCriteria.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+import org.json_voltpatches.JSONException;
+import org.json_voltpatches.JSONObject;
+import org.voltdb.StartAction;
+import org.voltdb.common.NodeState;
+
+import com.google_voltpatches.common.base.Throwables;
+import com.google_voltpatches.common.collect.ImmutableList;
+
+public class HostCriteria {
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/HostCriteria.java]: Can you document this class and its purpose?

File: src/frontend/org/voltcore/messaging/HostCriteria.java [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+                && jo.has(MESH_HASH) && jo.getString(MESH_HASH) != null && !jo.getString(MESH_HASH).trim().isEmpty()
+                && jo.has(START_ACTION) && jo.getString(START_ACTION) != null && !jo.getString(START_ACTION).trim().isEmpty()
+                && jo.has(NODE_STATE) && jo.getString(NODE_STATE) != null && !jo.getString(NODE_STATE).trim().isEmpty()
+                && jo.has(HOST_COUNT) && jo.getInt(HOST_COUNT) > 0;
+        } catch (JSONException e) {
+            return false;
+        }
+    }
+
+    protected final boolean paused;
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/HostCriteria.java]: member names with m_

File: src/frontend/org/voltcore/messaging/HostCriteria.java [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+                && jo.has(IS_BARE) && (jo.getBoolean(IS_BARE) ? true : true)
+                && jo.has(IS_ENTERPRISE) && (jo.getBoolean(IS_ENTERPRISE) ? true : true)
+                && jo.has(ADD_ALLOWED) && (jo.getBoolean(ADD_ALLOWED) ? true : true)
+                && jo.has(CONFIG_HASH) && jo.getString(CONFIG_HASH) != null && !jo.getString(CONFIG_HASH).trim().isEmpty()
+                && jo.has(MESH_HASH) && jo.getString(MESH_HASH) != null && !jo.getString(MESH_HASH).trim().isEmpty()
+                && jo.has(START_ACTION) && jo.getString(START_ACTION) != null && !jo.getString(START_ACTION).trim().isEmpty()
+                && jo.has(NODE_STATE) && jo.getString(NODE_STATE) != null && !jo.getString(NODE_STATE).trim().isEmpty()
+                && jo.has(HOST_COUNT) && jo.getInt(HOST_COUNT) > 0;
+        } catch (JSONException e) {
+            return false;
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/HostCriteria.java]: Need to log here or is it expected to get here?

File: src/frontend/org/voltcore/messaging/HostCriteria.java [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+        if (this == obj)
+            return true;
+        if (obj == null)
+            return false;
+        if (getClass() != obj.getClass())
+            return false;
+        HostCriteria other = (HostCriteria) obj;
+        if (bare != other.bare)
+            return false;
+        if (configHash == null) {
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/HostCriteria.java]: Document what two equal HostCriteria means.

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 7]
```diff
@@ -140,13 +141,43 @@ public Config(String coordIp, int coordPort) {
...
+        public static List<Config> generate(PortGenerator ports, int hostCount) {
+            checkArgument(ports != null, "port generator is null");
+            checkArgument(hostCount > 0, "host count %s is not greater than 0", hostCount);
+
+            ImmutableList.Builder<Config> lbld = ImmutableList.builder();
+            String [] coordinators = new String[hostCount];
+
+            for (int i = 0; i < hostCount; ++i) {
+                Config cnf = new Config(null, DEFAULT_INTERNAL_PORT);
+                cnf.zkInterface = "127.0.0.1:" + ports.next();
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: use zkinterface coming from cli

File: lib/python/voltcli/voltdb.d/init.py (hunk 1 above)
Comment (vtkstef) [inline lib/python/voltcli/voltdb.d/init.py]: not an easy one I'll take a look at it

File: src/frontend/org/voltcore/messaging/HostCriteria.java (hunk 2 above)
Comment (vtkstef) [inline src/frontend/org/voltcore/messaging/HostCriteria.java]: it is mostly there for documentation. it is not called anywhere but I left there intentionally to document the fields and their types

File: src/frontend/org/voltcore/messaging/HostCriteria.java (hunk 4 above)
Comment (vtkstef) [inline src/frontend/org/voltcore/messaging/HostCriteria.java]: this is a dto that offshoots from JoinerCriteria and is passed around in the initial mesh requests. all fields are immutable

File: src/frontend/org/voltcore/messaging/HostCriteria.java (hunk 5 above)
Comment (vtkstef) [inline src/frontend/org/voltcore/messaging/HostCriteria.java]: this is a check. if we get a json exception here is because the type of the fields mismatches. hence the hasCriteria check fails

File: src/frontend/org/voltcore/messaging/HostMessenger.java (hunk 7 above)
Comment (vtkstef) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: this generate is used only during testing. hence the cli had no bearing here

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 8]
```diff
@@ -176,6 +207,12 @@ public String toString() {
                 js.key("deadhosttimeout").value(deadHostTimeout);
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: make these constants...chasing becomes easy.

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 9]
```diff
@@ -668,34 +788,88 @@ private void prepSocketChannel(SocketChannel sc) {
...
+            m_foreignHosts = ImmutableMap.<Integer, ForeignHost>builder()
+                    .putAll(Maps.filterKeys(m_foreignHosts, not(equalTo(hostId))))
+                    .build();
         }
         if (fh != null) {
             fh.close();
         }
     }
 
+    public static class PleaDecision {
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: you need this public?

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 10]
```diff
@@ -936,6 +1120,130 @@ public void notifyOfHosts(
...
+                .filter(h -> h.getNodeState().operational())
+                .map(h -> h.getHostCount())
+                .findAny().orElse(m_criteria.getHostCount());
+
+        final int ksafety = m_criteria.getkFactor() + 1;
+        if (m_hostCriteria.size() < hostCount) {
+            return;
+        }
+        if (hostCount < m_criteria.getHostCount() && m_hostCriteria.size() <= hostCount) {
+            return;
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: Who retries? maybe its not obvious but could help on documentation somewhere. Can you document a whole mesh in javadoc.

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 11]
```diff
@@ -936,6 +1120,130 @@ public void notifyOfHosts(
...
+            m_paused.set(paused);
+        }
+
+        if (m_criteria.getStartAction() != StartAction.PROBE) {
+            m_probedDetermination.set(new Determination(
+                    m_criteria.getStartAction(), m_criteria.getHostCount()));
+            return;
+        }
+
+        StartAction determination = m_criteria.isBare() ?
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: What happens when some say bare and some dont?

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 12]
```diff
@@ -936,6 +1120,130 @@ public void notifyOfHosts(
...
+        } else if (operational > 0 && operational == hostCount) {
+            if (m_criteria.isAddAllowed()) {
+                hostCount = hostCount + ksafety;
+                determination = StartAction.JOIN;
+            } else {
+                try { shutdown(); } catch (Exception ignoreIt) {}
+                org.voltdb.VoltDB.crashLocalVoltDB("Node is not allowed to rejoin an already whole cluster");
+                return;
+            }
+        } else if (operational == 0 && bare == unmeshed) {
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: What if local is bare but others are not and cluster is not full?

Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: There needs to be some kind of trace logging in this whole process so that we can get details from field.

File: src/frontend/org/voltcore/messaging/JoinerCriteria.java [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+        }
+
+        // Try to interpret the specifier as an IP address.  Note we build
+        // the address rather than using the .is* methods because we want to
+        // use InetAddresses.toUriString to convert the result to a string in
+        // canonical form.
+        InetAddress addr = null;
+        try {
+          addr = InetAddresses.forString(host);
+        } catch (IllegalArgumentException e) {
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/JoinerCriteria.java]: log?

File: src/frontend/org/voltcore/messaging/JoinerCriteria.java [hunk 14]
```diff
@@ -0,0 +1,462 @@
...
+                m_meshHash,
+                m_enterprise,
+                m_startAction,
+                m_bare,
+                m_hostCount,
+                m_nodeStateSupplier.get(),
+                m_addAllowed
+                );
+    }
+    @Override
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/JoinerCriteria.java]: need an empty line above.

File: src/frontend/org/voltcore/messaging/JoinerCriteria.java [hunk 15]
```diff
@@ -0,0 +1,462 @@
...
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + (m_paused ? 1231 : 1237);
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/JoinerCriteria.java]: make 1231 and 1237 as finals easy to change.

File: src/frontend/org/voltcore/messaging/JoinerCriteria.java [hunk 16]
```diff
@@ -0,0 +1,462 @@
...
+    }
+
+    public void appendTo(JSONStringer js) throws JSONException {
+        js.object();
+        js.key("coordinators").array();
+        for (String coordinator: m_coordinators) {
+            js.value(coordinator);
+        }
+        js.endArray();
+        js.key("enterprise").value(m_enterprise);
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/JoinerCriteria.java]: make these constants.

File: src/frontend/org/voltcore/messaging/SocketJoiner.java [hunk 17]
```diff
@@ -66,24 +70,39 @@
 public class SocketJoiner {
 
     private static final int MAX_CLOCKSKEW = Integer.getInteger("MAX_CLOCKSKEW", 200);
+    private static final int RETRY_INTERVAL = Integer.getInteger("MESH_JOIN_RETRY_INTERVAL", 10);
+    private static final int RETRY_INTERVAL_SALT = Integer.getInteger("MESH_JOIN_RETRY_INTERVAL_SALT", 30);
     private static final int CRITICAL_CLOCKSKEW = 100;
```
Comment (akhanzode) [inline src/frontend/org/voltcore/messaging/SocketJoiner.java]: Make this property as well?

File: tests/frontend/org/voltcore/messaging/TestHostMessenger.java [hunk 18]
```diff
@@ -25,106 +25,1145 @@
...
-                try {
-                    hm3.start(null);
-                } catch (Exception e) {
-                    e.printStackTrace();
-                    exception.set(e);
-                }
-            }
-        };
+        HostMessengerThread hm2Start = new HostMessengerThread(hm2, exception);
+        HostMessengerThread hm3Start = new HostMessengerThread(hm3, exception);
```
Comment (akhanzode) [inline tests/frontend/org/voltcore/messaging/TestHostMessenger.java]: Do you have any legacy command check tests here? If not we should add

File: examples/adperformance/run.sh [hunk 19]
```diff
@@ -63,11 +63,15 @@ function jars-ifneeded() {
     fi
 }
 
+# Init to directory voltdbroot
+function voltinit-ifneeded() {
+    voltdb init --force
```
Comment (jhugg) [inline examples/adperformance/run.sh]: Why make a separate target that only has one line?

File: examples/voter/run.sh [hunk 20]
```diff
@@ -53,11 +53,16 @@ function jars-ifneeded() {
...
+function voltinit-ifneeded() {
+    voltdb init --force
+}
+
 # run the voltdb server locally
 function server() {
-    # note: "create --force" will delete any existing data
-    # use "recover" to start from an existing voltdbroot folder with data
-    voltdb create --force -H $STARTUPLEADERHOST
+    jars-ifneeded
```
Comment (jhugg) [inline examples/voter/run.sh]: Why is this line here?

File: lib/python/voltcli/verbs.py [hunk 21]
```diff
@@ -415,6 +416,7 @@ def __init__(self, subcommand,
         self.check_environment_config = check_environment_config
         self.force_voltdb_create = force_voltdb_create
         self.supports_paused = supports_paused
+        self.is_legacy_verb= is_legacy_verb
```
Comment (jhugg) [inline lib/python/voltcli/verbs.py]: Somewhere in verb.py we should add a comment explaining what "legacy" is. It can be short, hopefully with a pointer to more info, like a ticket number or something.

File: src/frontend/org/voltcore/messaging/HostCriteria.java (hunk 3 above)
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/HostCriteria.java]: Any new top-level class should have a comment explaining WHAT, WHY, and HOW (at least).

File: src/frontend/org/voltcore/messaging/HostCriteria.java (hunk 2 above)
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/HostCriteria.java]: When it fails, are you able to tell why? This kind of thing is typically harder to debug.

File: src/frontend/org/voltcore/messaging/HostCriteria.java [hunk 22]
```diff
@@ -0,0 +1,253 @@
...
+        this.meshHash = meshHash;
+        this.enterprise = enterprise;
+        this.startAction = startAction;
+        this.bare = bare;
+        this.hostCount = hostCount;
+        this.nodeState = nodeState;
+        this.addAllowed = addAllowed;
+    }
+
+    public boolean isPaused() {
```
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/HostCriteria.java]: If these are immutable, do we need all the accessors?

File: src/frontend/org/voltcore/messaging/HostCriteria.java [hunk 23]
```diff
@@ -0,0 +1,253 @@
...
+        }
+        return jo;
+    }
+
+    public boolean isUndefined() {
+        return UNDEFINED.equals(configHash) && UNDEFINED.equals(meshHash);
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/HostCriteria.java]: Is there a good reason this needs to be implemented? Does anyone call this? Can we just make it throw unsupported?

File: src/frontend/org/voltcore/messaging/HostCriteria.java [hunk 24]
```diff
@@ -0,0 +1,253 @@
...
+            ilb.add("Mismatched host parameters given at database startup");
+        }
+        if (!configHash.equals(o.configHash)) {
+            ilb.add("Mismatched deployment configuration given at database startup");
+        }
+        return ilb.build();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/HostCriteria.java]: This method doesn't follow our code conventions. Use {} for ifs or put the statement on the same line if it really makes sense to (it might here).

File: src/frontend/org/voltcore/messaging/HostMessenger.java (hunk 7 above)
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: When something in non-test code is only used for testing, COMMENT IT AS SUCH.

File: examples/adperformance/run.sh (hunk 19 above)
Comment (akhanzode) [inline examples/adperformance/run.sh]: I believe I removed this from commits/pr examples are still using old syntax...same comment for voter run.sh they should be as is on master.

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 25]
```diff
@@ -589,6 +658,57 @@ public void start(String request) throws Exception {
...
+            if (rejoiningHost == -1) {
+                return new PleaDecision(null, true, false);
+            } else {
+                String msg = "Only one host can rejoin at a time. Host "
+                      + rejoiningHost + " is still rejoining.";
+                return new PleaDecision(msg, false, true);
+            }
+        }
+        return new PleaDecision(null, true, false);
+    }
```
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: Soo.... comments?

cropped-this-is-fine-meme

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 26]
```diff
@@ -668,34 +788,88 @@ private void prepSocketChannel(SocketChannel sc) {
...
-            b.put(hostId, fh);
-            m_foreignHosts = b.build();
+            m_foreignHosts = ImmutableMap.<Integer, ForeignHost>builder()
+                    .putAll(m_foreignHosts)
+                    .put(hostId, fh)
+                    .build();
+        }
+    }
+
+    static final Predicate<Integer> in(final Set<Integer> set) {
```
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: Could these methods be generic-ized and put int some immutable map util? Are they not already somewhere?

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 27]
```diff
@@ -936,6 +1120,130 @@ public void notifyOfHosts(
...
 
+    public static class Determination {
+        public final StartAction startAction;
+        public final int hostCount;
+        private Determination(StartAction startAction, int hostCount) {
+            this.startAction = startAction;
+            this.hostCount = hostCount;
+        }
+        @Override
+        public int hashCode() {
```
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: Same comment about hash codes as before. Is it necessary?

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 28]
```diff
@@ -936,6 +1120,130 @@ public void notifyOfHosts(
...
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + hostCount;
+            result = prime * result
+                    + ((startAction == null) ? 0 : startAction.hashCode());
+            return result;
+        }
+        @Override
+        public boolean equals(Object obj) {
```
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: Same comment as before about if {} coding style.

File: lib/python/voltcli/verbs.py (hunk 21 above)
Comment (vtkstef) [inline lib/python/voltcli/verbs.py]: ok

File: src/frontend/org/voltcore/messaging/JoinerCriteria.java [hunk 29]
```diff
@@ -0,0 +1,462 @@
...
+        }
+
+        if (addr != null) {
+          return true;
+        }
+        // It is not any kind of IP address; must be a domain name or invalid.
+        return InternetDomainName.isValid(host);
+    }
+
+    protected final ImmutableSortedSet<String> m_coordinators;
```
Comment (jhugg) [inline src/frontend/org/voltcore/messaging/JoinerCriteria.java]: Why are these m_ when the members of HostCriteria are not?

Also not sold here on tons of accessors for final vars.

File: src/frontend/org/voltcore/messaging/HostCriteria.java (hunk 23 above)
Comment (vtkstef) [inline src/frontend/org/voltcore/messaging/HostCriteria.java]: Pretty easy to generate (eclipse -> Source -> generate hash and equals) It is good practice to implement them always particularly if instances of HostCriteria are part of sets

File: src/frontend/org/voltcore/messaging/HostMessenger.java (hunk 7 above)
Comment (vtkstef) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: ok

File: src/frontend/org/voltcore/messaging/HostMessenger.java (hunk 25 above)
Comment (vtkstef) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: ok

File: src/frontend/org/voltcore/messaging/HostMessenger.java (hunk 26 above)
Comment (vtkstef) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: not that I know. I just followed guava immutable map handling best practices

**Comment by vtkstef** [comment]
removing this pull request as the branch was erroneously with merge --squash. all comments so far have been addressed in  0fe09016f54d4b9fa103e57bb80fa5c4897ba5e1 and 23aeb98cebb8228ffd49ba2885f94b61267dac33
